header/p2p/session.go

```diff
@@ -0,0 +1,224 @@
+	headersPerPeer uint64 = 64
+)
+
+// session aims to divide a range of headers
+// into several smaller requests among different peers.
+type session struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	host       host.Host
+	protocolID protocol.ID
+	// peerTracker contains discovered peers with records that describes their activity.
+	queue *peerQueue
+
+	reqCh chan *p2p_pb.ExtendedHeaderRequest
+	errCh chan error
+}
+
+func newSession(ctx context.Context, h host.Host, peerTracker []*peerStat, protocolID protocol.ID) *session {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*2)
+	return &session{
+		ctx:        ctx,
+		cancel:     cancel,
+		protocolID: protocolID,
+		host:       h,
+		queue:      newPeerQueue(peerTracker),
+		errCh:      make(chan error),
+	}
+}
+
+// doRequest chooses the best peer to fetch headers and sends a request in range of available maxRetryAttempts
+func (s *session) doRequest(
+	ctx context.Context,
+	stat *peerStat,
+	req *p2p_pb.ExtendedHeaderRequest,
+	headers chan []*header.ExtendedHeader,
+) {
+	select {
+	case <-ctx.Done():
+		return
+	case <-s.ctx.Done():
+		return
+	default:
+	}
+	r, size, duration, err := s.requestHeaders(ctx, stat.peerID, req)
+	if err != nil {
+		if err == context.Canceled {
+			return
+		}
+		log.Errorw("requesting headers from peer failed."+
+			"Retrying the request from different peer", "failed peer", stat.peerID, "err", err)
+		s.reqCh <- req
+		return
+	}
+
+	h, err := s.processResponse(r)
+	if err != nil {
+		s.errCh <- err
+		return
+	}
+	log.Debugw("request headers from peer succeed", "from", s.host.ID(), "pID", stat.peerID, "amount", req.Amount)
+	// send headers to the channel, update peer stats and return peer to the queue, so it can be re-used in case
+	// if there are other requests awaiting
+	headers <- h
+	stat.updateStats(size, duration)
+	s.queue.push(stat)
+}
+
+// handleOutgoingRequest pops peer from the queue and sends a prepared request to the peer.
+func (s *session) handleOutgoingRequest(ctx context.Context, result chan []*header.ExtendedHeader) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-s.ctx.Done():
+			return
+		case req := <-s.reqCh:
+			stats := s.queue.calculateBestPeer(ctx)
+			go s.doRequest(ctx, stats, req, result)
+		}
+	}
+}
+
+// GetRangeByHeight requests headers from different peers.
+func (s *session) getRangeByHeight(ctx context.Context, from, amount uint64) ([]*header.ExtendedHeader, error) {
+	log.Debugw("requesting headers", "from", from, "to", from+amount)
+	requests := prepareRequests(from, amount, headersPerPeer)
+	result := make(chan []*header.ExtendedHeader, len(requests))
+	headers := make([]*header.ExtendedHeader, 0, amount)
+	s.reqCh = make(chan *p2p_pb.ExtendedHeaderRequest, len(requests))
+
+	go s.handleOutgoingRequest(ctx, result)
+	for _, req := range requests {
+		s.reqCh <- req
+	}
+
+	for i := 0; i < cap(result); i++ {
```

Comment:
readability nit:
```suggestion
	for i := 0; i < len(requests); i++ {
```

I see where you are coming from, but it's more readable to make all length dependent on one var.

Reply:
I thought we have to iterate over the capacity of channel and not the length. We can have len = 3, while cap = 8. So we will iterate 3 times and stop

Reply:
cap(result) == len(requests)

Reply:
requests are no the chan

Reply:
The nit still stands